### PR TITLE
sprint 7 일부 PR

### PR DESCRIPTION
**Pull request by ssaru**
1. preface
1-1. author
1-2. revision
1-3. table of contents

etc..

### CODE REVIEW

File: posts/03_01_dataloader.md [hunk 1]
```diff
@@ -39,7 +39,13 @@ class Dataset(object):
 
 
 
-python에서 `__getitem__`, `__len__`, `__add__`에 대해서 이야기하면 더 깊이 이야기할 수 있겠지만, 여기에서는 간단하게 정의하고 지나가겠습니다.
+일반적으로 이러한 pytorch의 Dataset 클래스를 상속받아, `__getitem__`, `__len__`을 overriding해서 필요한 모델에 맞는 커스텀 Dataset 클래스를 만들어서 사용하게 됩니다.
```
Comment (insurgent92) [inline posts/03_01_dataloader.md]:
```suggestion
일반적으로 이렇게 PyTorch의 Dataset 클래스를 상속받아 커스텀 Dataset 클래스를 만들고 `__getitem__`, `__len__`을 overriding해서 사용합니다.
```

File: posts/03_01_dataloader.md [hunk 2]
```diff
@@ -39,7 +39,13 @@ class Dataset(object):
...
+
+
+
+자세한 내용은 [Pytorch Tutorial](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html)을 확인해보시면 됩니다.
```
Comment (insurgent92) [inline posts/03_01_dataloader.md]:
```suggestion
자세한 내용은 [Pytorch Tutorial](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html)을 확인해 보시기 바랍니다. 
```

File: posts/03_01_dataloader.md [hunk 3]
```diff
@@ -48,11 +54,11 @@ python에서 `__getitem__`, `__len__`, `__add__`에 대해서 이야기하면 
 
 추상클래스인 `Dataset`은 최소한  `__getitem__`, `__len__` 함수 구현을 요구하고 있으므로, 우리는 이를 Object Detection에 맞춰서 Dataset 클래스를 상속받은 `VOC`클래스를 만들어봅시다.
```
Comment (insurgent92) [inline posts/03_01_dataloader.md]:
```suggestion
추상클래스인 `Dataset`은 최소한  `__getitem__`, `__len__` 함수 구현을 요구합니다. 그럼 이제 Dataset 클래스를 상속받아 Object Detection에 적합한 `VOC`클래스를 만들어봅시다.
```

File: posts/03_01_dataloader.md [hunk 4]
```diff
@@ -48,11 +54,11 @@ python에서 `__getitem__`, `__len__`, `__add__`에 대해서 이야기하면 
...
 
-
+​    
 
 ## VOC class
 
-VOC class는 앞에 설명한 `convert2Yolo`프로젝트를 같이 이용해서 구현됩니다. VOC class는 추상클래스인 `Dataset`에서 Object Detection data parsing  / 파일 존재 여부에 대한 로직이 추가되며, 추상클래스에서 요구하는 `__getitem__`, `__len__`함수를 구현합니다.
+VOC class는 앞에 설명한 `convert2Yolo`프로젝트를 같이 이용해서 구현됩니다. VOC class는 추상클래스인 `Dataset`에서 Object Detection data parsing  파일 존재 여부를 확인하는 `_check_exists()`함수에 대한 구현이 추가되며, 추상클래스에서 요구하는 `__getitem__`, `__len__`함수를 overriding하여 내부 함수를 구현합니다.
```
Comment (insurgent92) [inline posts/03_01_dataloader.md]:
```suggestion
VOC class 구현 시, 앞서 설명한 `convert2Yolo`프로젝트를 이용합니다. VOC class에 구현된 `_check_exists()`는 추상클래스인 `Dataset`에서 Object Detection data parsing  파일 존재 여부를 확인하는 함수이며, `__getitem__()`, `__len__()`는 추상클래스인 Dataset이 요구하는 함수를 overriding하여 구현한 함수입니다.
```

File: posts/04_02_Model.md [hunk 5]
```diff
@@ -0,0 +1,329 @@
+# 02) Model
+
+이번장에서는 YOLO 모델 구조 및 Objective function에 대해서 설명하고, 해당 내용을 Pytorch 코드레벨로 설명드리도록 하겠습니다.
```
Comment (insurgent92) [inline posts/04_02_Model.md]:
```suggestion
이번 장에서는 YOLO 모델 구조 및 objective function을 설명하고, 해당 내용을 PyTorch 코드레벨로 설명하겠습니다.
```

File: posts/04_02_Model.md [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+이전의 논문리뷰에서도 봤듯이 YOLO의 모델 구조는 다음과 같습니다.
```
Comment (insurgent92) [inline posts/04_02_Model.md]:
```suggestion
앞서 살펴보았듯이 YOLO의 모델 구조는 다음과 같습니다.
```

File: posts/04_02_Model.md [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+
+
+![YOLO model](https://user-images.githubusercontent.com/13328380/48620356-3290d280-e9e3-11e8-8488-0d17e57da49b.PNG)
+
+​    
+
+## YOLO class
+
+YOLO의 원 코드인 [Darknet](https://github.com/pjreddie/darknet)의 [yolov1.cfg](https://github.com/pjreddie/darknet/blob/master/cfg/yolov1.cfg) 파일을 확인하는 것을 통해서 YOLO의 pytorch model은 다음과 같이 코드로 구현할 수 있습니다. (완전히 똑같지는 않으나, 유사한 모델설계를 했습니다.)
```
Comment (insurgent92) [inline posts/04_02_Model.md]:
```suggestion
YOLO의 원 코드인 [Darknet](https://github.com/pjreddie/darknet)의 [yolov1.cfg](https://github.com/pjreddie/darknet/blob/master/cfg/yolov1.cfg) 를 살펴보면서 YOLO의 PyTorch model을 다음과 같이 구현하였습니다. (완전히 동일하지는 않으나, 유사하게 모델 설계를 했습니다)
```

File: posts/04_02_Model.md [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+            nn.LeakyReLU(),
+            nn.Dropout(self.dropout_prop)
+        )
+
+        self.fc2 = nn.Sequential(
+            nn.Linear(4096, 7 * 7 * ((10) + self.num_classes))
+        )
+```
+
+- 초기화할 때, `dropout`, `number of class`에 대한 파라미터를 받는 부분 빼고는 일반적인 YOLO model 형태로 작성하였습니다.
```
Comment (insurgent92) [inline posts/04_02_Model.md]:
```suggestion
- 초기화 시, `dropout`, `number of class` 파라미터 입력을 제외하면 일반적인 YOLO model 형태입니다. 
```

File: posts/04_02_Model.md [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+- 중요한 부분은 마지막의 `fc2`부분에서 `7 * 7 * ((10) + self.num_classes)`부분입니다. 해당 부분을 통해서 최종으로 7 x 7 x 30 block tensor가 나오게 되고 이를 이용하여 Detection에 필요한 요소들을 뽑아서 학습하거나 결과값을 이용하여 Detection Box를 그리기 때문입니다.
```
Comment (insurgent92) [inline posts/04_02_Model.md]:
```suggestion
- 중요한 부분은 모델의 마지막 `fc2`의 `7 * 7 * ((10) + self.num_classes)`입니다. 이 부분은 모델의 최종 출력으로 7 x 7 x 30 block tensor이고, 이를 이용하여 Detection에 필요한 요소들을 뽑아서 학습하거나 결과 값을 이용하여 Detection Box를 그리기 때문입니다.
```

Comment (insurgent92) [inline posts/04_02_Model.md]: 마지막 'fc2'가 왜 중요한지 보강이 필요할 것 같습니다.

File: posts/04_02_Model.md [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+
+  (여기서 10은 2개의 box값을 의미합니다. : 1 box \rightarrow `objectness`, `point_x_shift`, `point_y_shift`, `width_ratio`,`height_ratio`)
```
Comment (insurgent92) [inline posts/04_02_Model.md]:
```suggestion
  (여기에서 10은 2개의 box값을 의미합니다. : 1 box (`objectness`, `point_x_shift`, `point_y_shift`, `width_ratio`,`height_ratio`)
```

File: posts/04_02_Model.md [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+    elif isinstance(m, nn.BatchNorm2d):
+        nn.init.constant_(m.weight, 1)
+        nn.init.constant_(m.bias, 0)
+```
+
+​    
+
+## Forward
+
+다음과 같이 YOLO model을 작성했다면, Forward 함수는 다음과 같이 작성합니다.
```
Comment (insurgent92) [inline posts/04_02_Model.md]:
```suggestion
YOLO model 설계가 완료되었다면, Forward() 함수를 작성합니다.
```

File: posts/04_02_Model.md [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+
+![loss concept](https://user-images.githubusercontent.com/13328380/48621237-3c680500-e9e6-11e8-9e25-e256192c1648.png)
+
+
+
+`objectiveness map`이 나오면, 이를 반전시켜서 `non-objectiveness map`을 생성한 다음에 필요한 조건에 맞춰서 계산한 `Loss`를 계산해 더해줬습니다.
+
+
+
+### size loss 이슈
```
Comment (insurgent92) [inline posts/04_02_Model.md]: issue와 관련된 사항은 새로운 항목으로 빼도 좋을 것 같습니다. (가령 03). Issues) 
우리가 실제로 구현하면서 얻은 팁이니 흥미롭게 보는 독자들이 많을 것 같습니다.

File: posts/00_00_preface.md [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+**작성계획**
+
+2018년 8월부터 2019년 6월까지 10개월동안 Detection 메인스트림에 있는 YOLOv1, YOLOv2, YOLOv3, SSD, Faster R-CNN 모델을 구현하고 매 모델이 수렴될 때마다, 한 모델의 구현에 대해서 초안이 작성 된 뒤, 1~2주간의 내부리뷰를 거치며, 해당 리뷰를 통과한 자료에 한해서 점진적으로 공개됩니다.
+
+​    
+
+**이 문서의 활용 방법**
+
+이 문서의 전반적인 내용은 논문 위주와 모델 구현체의 파편을 참고하여 구현이 되었습니다. 본 문서는 해당 논문들을 레퍼런스하면서 공부하시면 좋습니다.
```
[the remark this answers is not in the document]
Comment (ssaru) [reply posts/00_00_preface.md]: 어떤 방향으로 보완이 필요할까요?

File: posts/04_02_Model.md (hunk 12 above)
Comment (ssaru) [reply posts/04_02_Model.md]: 좋습니다!

File: posts/02_02_Convert2Yolo.md [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+- UDACITY
+- KITTI
+
+
+
+해당 데이터셋이 아닌 다른 데이터셋을 사용하려면 직접 파싱코드를 작성하는 방법, [convert2Yolo](https://github.com/ssaru/convert2Yolo)에 해당 데이터셋을 파싱하는 코드를 작성, 새로운 툴을 사용하는 방법으로 이를 해결해야합니다.
+
+​    
+
+### 의존성 패키지 설치
```
[the remark this answers is not in the document]
Comment (ssaru) [reply posts/02_02_Convert2Yolo.md]: 도식화라는게 어떤 것을 의마하나요?

File: posts/04_02_Model.md (hunk 9 above)
Comment (ssaru) [reply posts/04_02_Model.md]: 어떤게 있을까요?

File: posts/00_00_preface.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (insurgent92) [reply posts/00_00_preface.md]: 활용 방법이라는 제목인데
문서의 속성(논문 + 구현)과 
이 문서를 읽을 때 해당 논문을 참고해서 공부하면 좋다.
는 활용 방법이랑은 조금 다른 것 같아서요!

File: posts/04_02_Model.md (hunk 9 above)
Comment (insurgent92) [reply posts/04_02_Model.md]: 1. 
> 초기화할 때, `dropout`, `number of class`에 대한 파라미터를 받는 부분 빼고는 일반적인 YOLO model 형태로 작성하였습니다.   

초기화 시 일반적인 YOLO model의 형태와, dropout`, `number of class` 파라미터 부분이 일반적인 YOLO model의 형태를 벗어나게 하는건지에 설명이 더 필요할 것 같습니다. 

2.
>중요한 부분은 마지막의 `fc2`부분에서 `7 * 7 * ((10) + self.num_classes)`부분입니다. 해당 부분을 통해서 최종으로 7 x 7 x 30 block tensor가 나오게 되고 이를 이용하여 Detection에 필요한 요소들을 뽑아서 학습하거나 결과값을 이용하여 Detection Box를 그리기 때문입니다.

말씀하신 내용을 요약해보면 
'fc2'의 마지막부분이 중요한 이유가 model output이기 때문이다. 라고 이해했습니다. 
문맥을 조금 더 자연스럽게 만들면 좋을 것 같아요  

3. `detection Box`는 `bounding box`로 용어를 통일하면 좋을 것 같습니다.

File: posts/04_02_Model.md (hunk 12 above)
Comment (insurgent92) [reply posts/04_02_Model.md]: 좋습니다! 
이슈 추가했습니다.
https://github.com/DeepBaksuVision/Modu_ObjectDetection/issues/21

File: posts/00_00_preface.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (ssaru) [reply posts/00_00_preface.md]: 네! 해당 내용 보강해서 다시 리뷰하도록 하겠습니다.
